execution environments: add EE overview page

Andersson007 opened the pull request:
Fixes https://github.com/ansible-community/community-team/issues/232

execution environments:
- add EE overview page
- add build and test your first EE

next:
- i'll create the other docs one by one listed in the epic
- I'll integrate them with this overview and how-to

Andersson007 commented:
ready_for_review

Akasurde commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -0,0 +1,147 @@
+* **Ansible package**: you install it using ``pip install ansible`` or an OS distribution package manager; it provides ansible-core and a big set of Ansible collections in the *batteries included* manner.
+* **Ansible**: in the context of this document, is the Ansible package or ansible-core plus a set of collections installed on the Ansible control node.
+* **Ansible control node**: the machine from which you run Ansible.
+* **Containerization**: it's what you typically use ``podman`` or ``docker`` for.
```
```suggestion
* **Container Runtime**: it is what you typically use ``podman`` or ``docker`` for running containers.
```

Akasurde commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -0,0 +1,147 @@
+Why Execution Environments were introduced
+==========================================
+
+The Ansible automation execution environments were introduced to resolve the following issues
```
```suggestion
The Ansible Execution Environments were introduced to resolve the following issues
```

Akasurde commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -0,0 +1,147 @@
+The dependencies have to be installed and sometimes can conflict with each other.
+
+One way to **partly** resolve dependency issue is
+to use Python virtual environments on Ansible controle node.
```
```suggestion
to use Python virtual environments on Ansible control node.
```

Akasurde commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -0,0 +1,147 @@
+Each user works with a limited set of services and wants to use only ``ansible-core``
+and a corresponding set of Ansible collections automating those services.
+
+Each user is not interested in having someone else's content the users don't need including dependencies
```
```suggestion
Each user is not interested in having someone else's content the users do not need including dependencies
```

Akasurde commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -0,0 +1,147 @@
+* ansible-core
+* none or more Ansible collections
+* Python
+* dependencies
```
```suggestion
* Python and System dependencies 
```

Akasurde commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,181 @@
+
+  ---
+  collections:
+    - name: community.postgresql 
```
```suggestion
    - name: community.postgresql
```

Akasurde commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,181 @@
+
+  psycopg2-binary
+
+4. Create the ``execution-environment.yml`` definition file that specifies which content you want to include.
+List the files you created at previous steps:
+
+.. code-block:: yaml
+
+  ---
+  version: 1
+
+  dependencies:
+    galaxy: requirements.yml
+    python: requirements.txt
```
We have Ansible Builder version 3, let us use it. 

```yaml
---
version: 3
images:
  base_image:
    name: quay.io/centos/centos:stream9

dependencies:
  ansible_core:
    package_pip: ansible-core==2.14.4
  ansible_runner:
    package_pip: ansible-runner
  galaxy: requirements.yml
  python: requirements.txt
```

For more information - https://ansible-builder.readthedocs.io/en/stable/definition/#version-3-format

Andersson007 replied:
great, thanks for the link!

Akasurde commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,181 @@
+
+  $ mkdir demo
+
+Besides container artifacts, it will contain a ``Containerfile`` generated by ``ansbible-builder``,
```
```suggestion
Besides container artifacts, it will contain a ``Containerfile`` generated by ``ansible-builder``,
```

Akasurde commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,181 @@
+    galaxy: requirements.yml
+    python: requirements.txt
+
+5. Create the ``demo`` build context directory:
```
Not sure if we want to mention this step, since Ansible Builder automatically creates a context directory.

Andersson007 replied:
Yes, i know, decided to put it explicitly as, i think, it's something common that people want to create more than one EE

Akasurde commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,181 @@
+
+.. code-block:: bash
+
+  $ mkdir demo/inventory demo/project
```
```suggestion
  $ mkdir -p demo/{inventory,project}
```

Akasurde commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,181 @@
+
+If you use ``docker``:
+
+* add the ``--process-isolation-executable docker`` argument to the ``ansible-runner`` commands below
```
I would favor `ansible-navigator` instead of `ansible-runner` for EE demo purposes since it is easy to use.

Andersson007 replied:
thanks for the suggestion! I'm not yet familiar with navigatior. It's in the plan.
Let's leave it as is for now and I'll take using navigator here into consideration when learning it.

Andersson007 commented:
@Akasurde changing `version:` to 3 breaks the how to unfortunately. I've changed it back to 1.
The errors with 3 are:
```
$ ansible-runner run demo --playbook test_localhost.yml --container-image localhost/postgresql_ee --process-isolation
[WARNING]: Unable to parse /runner/inventory/hosts as an inventory source
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: error in 'jsonfile' cache plugin while trying to create cache dir
/runner/artifacts/4e81c492-2424-4dd7-ad42-9913185af4b1/fact_cache : b"[Errno
13] Permission denied:
'/runner/artifacts/4e81c492-2424-4dd7-ad42-9913185af4b1/fact_cache'"
[WARNING]: provided hosts list is empty, only localhost is available. Note that
the implicit localhost does not match 'all'
ERROR! Invalid callback for stdout specified: awx_display
```
Added the hosts file with as below, now there's another error
```
[student@fedora03 postgresql_ee]$ vim demo/inventory/hosts
[student@fedora03 postgresql_ee]$ ansible-runner run demo --playbook test_localhost.yml --container-image localhost/postgresql_ee --process-isolation
[WARNING]: error in 'jsonfile' cache plugin while trying to create cache dir
/runner/artifacts/e130ca40-a015-463c-9998-33bf9c2c3325/fact_cache : b"[Errno
13] Permission denied:
'/runner/artifacts/e130ca40-a015-463c-9998-33bf9c2c3325/fact_cache'"
ERROR! Invalid callback for stdout specified: awx_display
```
Changing `version:` to 1 fixes the issues

felixfontein commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,182 @@
+
+.. code-block:: bash
+
+  $ mkdir demo
```
I don't really understand why you want to create this directory manually, and why to use `--context` at all. Why not simply use the default `context` directory created and used by ansible-builder?

felixfontein commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,182 @@
+
+.. code-block:: bash
+
+  $ mkdir demo/{inventory,project}
```
In particular: why are you using the context directory for playbooks and inventory to run something in the EE? Running things in the EE should happen in a completely different environment (directory) than where the EE is in. These two steps should have nothing in common except that there exists a container image (the EE).

Using the same directory for two unrelated things is confusing, users might think it is important to use the same directory for these unrelated things.

Andersson007 replied:
>In particular: why are you using the context directory for playbooks and inventory to run something in the EE?

heh, i don't know why I'm using it TBH:)
probably because there's a lack of docs and it's just the way things work:)
so please just suggest better ways of doing things:)
i'll take another look at the runner's docs and  maybe at navigator's docs now as @Akasurde 's suggesting

felixfontein commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,182 @@
+
+.. code-block:: bash
+
+  $ ansible-runner run demo --playbook test_localhost.yml --container-image localhost/postgresql_ee --process-isolation
```
Why are you using ansible-runner and not ansible-navigator? I was assuming that ansible-runner is more like a low-level tool that is used by AWX and ansible-navigator, and that end-users should use ansible-navigator. (I'm not sure where I got this impression from though...)

Andersson007 replied:
> Why are you using ansible-runner and not ansible-navigator? I was assuming that ansible-runner is more like a low-level tool that is used by AWX and ansible-navigator, and that end-users should use ansible-navigator. (I'm not sure where I got this impression from though...)

Probably because i haven't read the navigator's docs yet;) It's in my list, i'll prioritize it, thanks

felixfontein commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,182 @@
+
+Here, we will test the EE you created in the :ref:`Building your first EE<build_first_ee>` section against the localhost and a remote target.
+
+You are still in the ``postgresql_ee`` directory.
```
I would change to another directory. The EE has been built, and the only thing you need from now on is the built container image.

Andersson007 replied:
ah, cool, thanks!

Andersson007 commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -10,7 +10,64 @@ Ansible Execution Environments Overview
+
+The dependencies have to be installed and sometimes can conflict with each other.
+
+Moreover, typically an administrator doesn't really need the whole Ansible package with those hundred collections. In most cases, and it's considered a good practice, we need to install only ``ansible-core`` plus a set of collections for specific tasks if needed, no more.
```
yeah, that's true:) thanks! i follow it but sometimes put the short forms automatically, haven't written the docs for ages:)

Andersson007 commented:
@felixfontein @Akasurde done, thanks!
your feedback is much appreciated, PTAL.
a couple of things:
- next thing i'll do is to figure out how to make it work w/o errors using version 3 of the execution-environment.yml file format
- i'll take a look at navigator docs

samccann commented:
@Andersson007 I'm unclear why most of this is here and not being added to the builder docs themselves? I guess I can understand the overview but the overview also exists in the builder docs.

cidrblock commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,156 @@
+
+  $ mkdir postgresql_ee && cd postgresql_ee
+
+2. Create the ``requirements.yml`` file which will specify collections you want to include in your EE.
```
I think we should recommend the single file approach here rather than multiple files.
 e.g. https://github.com/ansible/ansible-builder/blob/devel/demo/execution-environment.yml

Andersson007 replied:
yep, wasn't sure about best practices. it simplifies the guide, changed, thanks

cidrblock commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,156 @@
+Run against a remote target
+---------------------------
+
+1. Create the ``hosts`` file containing:
```
IMHO, we should generally encourage yaml style invetory files

Andersson007 replied:
sgtm, changed, thanks

cidrblock commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,156 @@
+Run against a remote target
+---------------------------
+
+1. Create the ``hosts`` file containing:
```
note, the inventory and playbook (and ansible.cfg file) should be in the project directory (typically the CWD for ease of use with navigator)

cidrblock commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -0,0 +1,148 @@
+* podman or docker
+* python-pip: to install the tools
+* ansible-builder: to build EEs
+* ansible-navigator: to run EEs
```
ansible-navigator depends on ansible-builder which could be noted here

Andersson007 replied:
do you think it'd harm if it's not here? I'm trying not to put any info that's not strictly necessary as, generally, users have to install all the mentioned packages in order to build and test EEs anyway

Andersson007 replied:
Ah, I see now, yep, makes sense to change, done, thanks!

cidrblock commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,156 @@
+
+  $ ansible-navigator run test_localhost.yml --execution-environment-image postgresql_ee --mode stdout --pull-policy missing
+
+The run will return facts gathered inside the container.
```
Maybe......  "You may notice the facts being gathered are about the container and not the developr machine.  This is because the ansible playbook was run inside the container"

Andersson007 replied:
done, thanks

cidrblock commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -0,0 +1,148 @@
+* ansible-navigator
+* Ansible AWX
+* Automation controller
+* ansible-runner
```
although maybe a bit more advanced, the ansible vscode extention along side the vscode container exention is another nice way to take adavantage of the EE https://code.visualstudio.com/docs/devcontainers/containers

cidrblock replied:
(in which case, the shell opens in the container and the tradtional ansible-playboo command can be used)

oraNod commented:
@Andersson007 That sounds fine to me. I'll add some more review comments but I don't think I'd classify the content as "How To". It seems more tutorial and conceptual to me, which I agree is specific to the builder docs and belongs there.

Andersson007 commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -0,0 +1,147 @@
+* none or more Ansible collections
+* Python
+* Python and system dependencies
+* custom user needs
```
Oh, you're right, great suggestion, i felt that something was wrong with that chapter

Andersson007 commented on execution_environments/how_to_build_and_test_first_ee.rst:
```diff
@@ -0,0 +1,164 @@
+  If a collection you want to install does not have the ``requirements.txt`` file,
+  you need to add Python dependencies specified in the collection's documentation to your the definition file manually.
+
+Read more about the EE definition file format in the `Ansible Builder documentation <https://ansible-builder.readthedocs.io/en/stable/definition/>`_.
```
I like this suggesting, always get something new, thanks!

Andersson007 commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -0,0 +1,147 @@
+* **Ansible package**: you install it using ``pip install ansible`` or an OS distribution package manager; it provides ansible-core and a big set of Ansible collections in the *batteries included* manner.
+* **Ansible**: in the context of this document, is the Ansible package or ``ansible-core`` plus a set of collections installed on the Ansible control node.
+* **Ansible control node**: the machine from which you run Ansible.
+* **Container Runtime**: it is what you typically use ``podman`` or ``docker`` for running containers.
```
SGTM, I added a short and very general definition. The more details explanation will come right below in the next section.

Andersson007 commented on execution_environments/execution_environments_overview.rst:
```diff
@@ -0,0 +1,147 @@
+.. _ee_rationale:
+
+Why Execution Environments were introduced
+==========================================
```
changed across the docs, thanks

Andersson007 commented:
@oraNod thanks for the great review!
All should now be incorporated, PTAL

oraNod commented:
@Andersson007 Hey I hope it's OK but I just pushed a commit with a bunch of suggested edits. Feel free to take them or leave them.

I guess there are a few things going on. First I think it's good to encourage using Python virtual environments as much as possible. That way users can just delete everything when they're done.

Another thing is dropping the dollar sign from commands. That just makes it easier to copy/paste directly from the codeblock in the HTML. Also consider using the cat EOF thing so users can copy/paste into their terminals to create the files. I feel like it's always better to try and allow users to copy/paste/run. It's faster and can avoid errors. Copying yaml from a code block to a file can lead to spacing or character issues that cause commands to fail.

I also moved a couple other things around. Happy to go over any of the other changes if you'd like me to. Cheers.

Andersson007 commented:
merging this, thanks everyone!
I'll submit PRs agains builder's docs

felixfontein commented:
Thanks everyone for helping to improve the EE docs!

Andersson007 commented:
@felixfontein we are currently deciding where this will live https://github.com/ansible-community/community-team/issues/243
your input would be much appreciated
